### PR TITLE
[Frontend]Adding loading animation for live Tweets feature

### PR DESCRIPTION
**Pull request by AC-Unicorn**
### background ###
Our live tweet feature sometimes taking 10+ seconds to retrieve data, 
In order to increase the user experience of this feature, we decide to add a loading animation
before tweets arrive. 

### Logistic ###
The animation will be triggered every time that the user triggers an event (drag, zoom, search, timebar), it will disappear when the first tweet arrives. The animation is purely implemented by CSS
animation, so it's pretty customizable.  There are a few options: 
image

New Feature:
after 20 seconds no tweets were showing, we will notify the status of live tweets. 
 

### Current Animation ###

Original animation:

image

after 20s without tweets coming:

image

### CODE REVIEW

**Comment by chenlica** [comment]
I don't have a strong preference.  How about the last one?

**Comment by AC-Unicorn** [comment]
> I don't have a strong preference. How about the last one?

Last one commonly used for video loading, that's why I don't like it in the first place.

**Comment by chenlica** [comment]
Any one should be good.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 1]
```diff
@@ -160,13 +163,27 @@ angular.module("cloudberry.sidebar", ["cloudberry.common"])
...
+          noSampleTweetsCount++;
+        }
+
         if($("#tweet").children().length > 20)
         {
           $("#tweet").children().last().remove();
         }
+
+        if($("#loadingAnime").length!=0 && noSampleTweetsCount>=10){
+          $("#loadingMsg").html("<p>Keyword maybe two rare, expecting longer time to see sample tweets</p>")
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]:
```suggestion
          $("#loadingMsg").html("<p>Keyword may be too rare, expecting longer time to see sample tweets</p>")
```

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 2]
```diff
@@ -160,13 +163,27 @@ angular.module("cloudberry.sidebar", ["cloudberry.common"])
     function startLiveTweetsConsumer() {
       $scope.liveTweetsConsumer = window.setInterval(function() {
         if (liveTweetsQueue.length > 0){
+          //reset the count since there is result
+          noSampleTweetsCount = 0;
+          if($("#loadingAnime").length!=0){
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]:
```suggestion
          if ($("#loadingAnime").length !== 0) {
```

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 3]
```diff
@@ -160,13 +163,27 @@ angular.module("cloudberry.sidebar", ["cloudberry.common"])
...
+            $("#tweet").html("");
+          }
           var data = liveTweetsQueue.pop();
           drawTweets(data);
         }
+        else{
+          noSampleTweetsCount++;
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]:
```suggestion
          noSampleTweetsCount ++;
```

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 4]
```diff
@@ -160,13 +163,27 @@ angular.module("cloudberry.sidebar", ["cloudberry.common"])
...
         if (liveTweetsQueue.length > 0){
+          //reset the count since there is result
+          noSampleTweetsCount = 0;
+          if($("#loadingAnime").length!=0){
+            $("#tweet").html("");
+          }
           var data = liveTweetsQueue.pop();
           drawTweets(data);
         }
+        else{
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]:
```suggestion
        else {
```

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 5]
```diff
@@ -160,13 +163,27 @@ angular.module("cloudberry.sidebar", ["cloudberry.common"])
...
+          noSampleTweetsCount++;
+        }
+
         if($("#tweet").children().length > 20)
         {
           $("#tweet").children().last().remove();
         }
+
+        if($("#loadingAnime").length!=0 && noSampleTweetsCount>=10){
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]:
```suggestion
        if ($("#loadingAnime").length !== 0 && noSampleTweetsCount >= 10) {
```

**Pull request merged**